### PR TITLE
feat(daemon): add raw_data endpoint

### PR DESCRIPTION
**Pull request by dysbulic**
This PR addresses #1394 and adds a `/raw_data` endpoint to the Ceramic daemon which returns the current `content` of a stream.

The purpose is to allow programs without knowledge of Ceramic to access the data in streams.

### CODE REVIEW

File: packages/cli/src/ceramic-daemon.ts
Comment (stbrody) [inline packages/cli/src/ceramic-daemon.ts:401]: Instead of this, I think it would make more sense to add a `raw_data` accessor to the [Stream type](https://github.com/ceramicnetwork/js-ceramic/blob/881d7f0f17de820290ba6b5b7f4b19e00d2eed6c/packages/common/src/stream.ts#L130).  The `raw_data` accessor can work very similarly to the [metadata accessor](https://github.com/ceramicnetwork/js-ceramic/blob/881d7f0f17de820290ba6b5b7f4b19e00d2eed6c/packages/common/src/stream.ts#L145), just returning the `content` field instead of the `metadata` field.

File: packages/cli/src/__tests__/ceramic-daemon.test.ts
Comment (stbrody) [inline packages/cli/src/__tests__/ceramic-daemon.test.ts:299]: Since you're not testing anything related to anchoring in this test, you can remove the calls to anchorDoc and add `{ anchor: false }` to the `create` and `update` calls, which will speed up the test

File: packages/cli/src/__tests__/ceramic-daemon.test.ts
Comment (stbrody) [inline packages/cli/src/__tests__/ceramic-daemon.test.ts:306]: It might be a more realistic test to use [fetchJson](https://github.com/ceramicnetwork/js-ceramic/blob/881d7f0f17de820290ba6b5b7f4b19e00d2eed6c/packages/common/src/utils/http-utils.ts#L9) to hit the HTTP endpoint rather than calling into the daemon object directly.

**Review by stbrody — CHANGES_REQUESTED** [review]
Basic approach looks good, a few suggestions inline.  Thanks for taking on the work to do this and make a PR yourself!

File: packages/cli/src/__tests__/ceramic-daemon.test.ts
Comment (dysbulic) [reply packages/cli/src/__tests__/ceramic-daemon.test.ts]: I tried [adding `anchor: false` to the various calls](https://github.com/ceramicnetwork/js-ceramic/pull/1395/files#diff-077b6994383a23f7d2210366626822554ebbcde4d9de8e8a5f94a93a7d1b1141R298-R300) and the test fails as through the updates are not applied:

```
● Ceramic interop: core <> http-client › can get stream contents from /raw_data

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 1

      Object {
    -   "test2": "def",
    +   "test": 123,
      }
```

Comment (stbrody) [reply packages/cli/src/__tests__/ceramic-daemon.test.ts]: Most likely because the current version lives in `state.next.content` not in `state.content`.  If you switch the raw_data endpoint to use the new `rawData` accessor you added to the Stream class, that should fix it.

**Comment by ukstv** [comment]
IMO, the endpoint is better be `/:stream_id/content`, part of `streamsRouter`.

**Comment by ukstv** [comment]
And +1 to @stbrody comments.

**Comment by dysbulic** [comment]
> IMO, the endpoint is better be `/:stream_id/content`, part of `streamsRouter`.

I thought that `content` made more sense for the name given the names of the elements of the objects. @oed said that is [too closely tied to existing stream types](https://discordapp.com/channels/682786569857662976/801539259353923655/843422531059056650), thus the name `raw_data` was used.

File: packages/cli/src/__tests__/ceramic-daemon.test.ts
Comment (dysbulic) [reply packages/cli/src/__tests__/ceramic-daemon.test.ts]: Done. I needed to [expose the port](https://github.com/dysbulic/js-ceramic/blob/feat/raw_data-endpoint/packages/cli/src/ceramic-daemon.ts#L142) in the daemon in order to know where to connect. I also [exposed the hostname](https://github.com/dysbulic/js-ceramic/blob/feat/raw_data-endpoint/packages/cli/src/ceramic-daemon.ts#L143) b/c the two generally go together and it looked lopsided w/ just one.

File: packages/cli/src/__tests__/ceramic-daemon.test.ts
Comment (dysbulic) [reply packages/cli/src/__tests__/ceramic-daemon.test.ts]: Done. Tests are passing now.

File: packages/cli/src/ceramic-daemon.ts
Comment (dysbulic) [reply packages/cli/src/ceramic-daemon.ts]: Done.

**Comment by oed** [comment]
> I thought that content made more sense for the name given the names of the elements of the objects. @oed said that is too closely tied to existing stream types, thus the name raw_data was used.

Yeah, `content` doesn't really make sense for Caip10Links for example. I think `only-data` might be more descriptive though!

**Comment by stbrody** [comment]
> > I thought that content made more sense for the name given the names of the elements of the objects. @oed said that is too closely tied to existing stream types, thus the name raw_data was used.
> 
> Yeah, `content` doesn't really make sense for Caip10Links for example. I think `/only-data` might be more descriptive though!

Personally I like `raw_data` better than `only_data`.  I like it because the important part in my mind isn't that we're getting the data instead of the metadata, it's that we're getting the data in the raw form from the state object, without any of the streamtype-specific presentation of that data that the more standard APIs provide. The object returned from `raw_data` might require some streamtype specific knowledge to know how to interpret/transform into something useful.

But I also don't feel super strongly.  If you prefer `only_data` @oed that's fine by me.

**Review by stbrody — APPROVED** [review]
LGTM

@oed, care to make a final ruling on `raw_data` vs `only_data` or something else?

**Comment by stbrody** [comment]
After a lengthy discussion this morning @oed and I decided to expose `content` as a primary accessor on the Stream type, and change this new endpoint to be at `/api/v0/streams/contents`.  I have gone ahead and updated this PR to incorporate those changes, as well as to update it to current develop branch.

@oed, can you take a final look at this PR and merge it if it looks good?

File: packages/cli/src/ceramic-daemon.ts
Comment (ukstv) [inline packages/cli/src/ceramic-daemon.ts:142]: @dysbulic Nope. Properties like this should be initialized in the constructor.

File: packages/cli/src/ceramic-daemon.ts
Comment (ukstv) [reply packages/cli/src/ceramic-daemon.ts:202]: Non-actionable, probably: +1 to the current `/:streamid/content`, +1 to `contents -> content` rename.

**Review by ukstv — APPROVED** [review]
Pls, @dysbulic fix initialisation of the properties, otherwise LTGM.

I would really, really love to have url endpoint as singular `content`, but I defer to @oed and @stbrody here. Having "content" in code and plural "contents" in url is a confusing to me.

File: packages/cli/src/__tests__/ceramic-daemon.test.ts
Comment (oed) [inline packages/cli/src/__tests__/ceramic-daemon.test.ts:293]:
```suggestion
    it('can get stream contents from /streams/content', async () => {
```

File: packages/cli/src/__tests__/ceramic-daemon.test.ts
Comment (oed) [inline packages/cli/src/__tests__/ceramic-daemon.test.ts:303]:
```suggestion
        `http://localhost:${daemon.port}/api/v0/streams/content/${doc.id}`
```

File: packages/cli/src/ceramic-daemon.ts
Comment (oed) [inline packages/cli/src/ceramic-daemon.ts:218]:
```suggestion
    contentsRouter.getAsync('/:streamid', this.content.bind(this))
```

File: packages/cli/src/ceramic-daemon.ts
Comment (oed) [inline packages/cli/src/ceramic-daemon.ts:400]:
```suggestion
  async content(req: Request, res: Response): Promise<void> {
```

**Review by oed — COMMENTED** [review]
These are just suggestions, need to be verified to be working.

**Comment by stbrody** [comment]
Updates based on last round of comments made.  Tests passed locally.  Back to you @oed for final approval

**Comment by dysbulic** [comment]
> Updates based on last round of comments made.

Thanks for doing this. I was getting a failing test in another part of the system and am happy to not have to figure out what's going on right now.

**Review by v-stickykeys — COMMENTED** [review]
I'd like to second @ukstv and propose that we generally aim for REST so that things are predictable/standardized rather than haphazard.

He already mentioned above but this would (1) use the streams router and add a path there instead of creating a new router and (2) change `/streams/content/<streamId>` to `/streams/<streamId>/content`.

Subtle but I think will be effective moving forward.

Additionally, want to check on the ordering of routes to ensure errors continue to propagate correctly.

File: packages/cli/src/ceramic-daemon.ts
Comment (stbrody) [reply packages/cli/src/ceramic-daemon.ts]: Done

File: packages/cli/src/ceramic-daemon.ts
Comment (stbrody) [reply packages/cli/src/ceramic-daemon.ts:211]: FWIW all the tests passed. I prefer this ordering as I think it's a little clearer to the reader what's happening as we go from more general things to more specific things.  But if it breaks things then of course we should change it.

Comment (v-stickykeys) [reply packages/cli/src/ceramic-daemon.ts:211]: Good point that we should have tests here. I can work on that this sprint as I test the metrics logs again

**Review by v-stickykeys — APPROVED** [review]
Confirmed that this works afaict! Nitpick if you don't mind @dysbulic 
let's move `baseRouter.use(errorHandler(this.diagnosticsLogger))` under `app.use('/api/v0', baseRouter)`

**Comment by oed** [comment]
Looks like this can be merged after a rebase 🎉

**Comment by stbrody** [comment]
Rebased and merged!  Thank you for your patience @dysbulic!

**Pull request merged**
